exodus_gw/worker/publish.py

```
@@ -1,87 +1,171 @@
+            if item_ids:
+                statement = select(Item).where(Item.id.in_(item_ids))
+                for partition in self.db.execute(statement).partitions():
+                    del_items = [row.Item for row in partition]
+                    write_batches(
+                        self.env, del_items, self.from_date, delete=True
+                    )
 
-            if not items_written or (last_items and not last_items_written):
-                items = items + last_items if last_items else items
-                write_batches(env, items, from_date, delete=True)
 
-                task.state = schemas.TaskStates.failed
-                publish.state = schemas.PublishStates.failed
-                db.commit()
-                return
-        except Exception:
-            LOG.exception("Task %s encountered an error", task.id)
+@dramatiq.actor(time_limit=Settings().actor_time_limit)
+def commit(publish_id: str, env: str, from_date: str) -> None:
+    actor_msg_id = CurrentMessage.get_current_message().message_id
+    commit_obj = Commit(publish_id, env, from_date, actor_msg_id)
 
-            task.state = schemas.TaskStates.failed
-            publish.state = schemas.PublishStates.failed
-            db.commit()
-            return
+    if not commit_obj.should_write():
+        return
 
-    task.state = schemas.TaskStates.complete
-    publish.state = schemas.PublishStates.committed
-    db.commit()
+    commit_obj.task.state = TaskStates.in_progress
+    commit_obj.db.commit()
+
+    try:
+        commit_obj.write_publish_items()
+        commit_obj.task.state = TaskStates.complete
+        commit_obj.publish.state = PublishStates.committed
+        commit_obj.db.commit()
+    except Exception as exc_info:
+        commit_obj.rollback_written_items(exc_info)
+        commit_obj.task.state = TaskStates.failed
+        commit_obj.publish.state = PublishStates.failed
+        commit_obj.db.commit()
+        raise
```

Hmm, I don't think we should raise here.  Note that if the actor raises, it's going to be retried by dramatiq a few times.  But that will definitely fail because the next try will see that the task state has already been moved to "failed" which will make should_write() return False.

I think the error handling logic needs to be like this:

- Either: do NOT mark task as failed, let exception be raised so we will be retried, and combine this with logic somewhere else to set task to failed if it is retried too many times.
- Or: mark task as failed and let the actor return successfully; implement desired retry logic inside of the actor.  This is the way it worked prior to this commit.

However, if you do drop the raise here, make sure to add appropriate logging.  As right now it seems that the exception won't be logged prior to raise if written_item_ids is empty.